E2E tests for Default Application Catalog

iakmc opened the pull request:
**What this PR does / why we need it**:
Prow jobs and E2E tests are added. There are prow jobs for each application from the default app catalog, and each job verifies that the ApplicationInstallation object has been installed successfully, its conditions are in a healthy state, a Helm release has been deployed, and the pods of the given application have been deployed successfully.
NOTE: The test for Falco fails, and it will be fixed in the [following issue](https://github.com/kubermatic/kubermatic/issues/14313).

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #https://github.com/kubermatic/kubermatic/issues/13934

**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/kind chore
-->

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Prow jobs and E2E tests are added. There are prow jobs for each application from the default app catalog, and each job verifies that the ApplicationInstallation object has been installed successfully, its conditions are in a healthy state, a Helm release has been deployed, and the pods of the given application have been deployed successfully. Additionally, end-to-end (E2E) tests are added to verify that the pods for every application in the default application catalog are deployed successfully. These tests help ensure compatibility when updating the version of each application.
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
NONE
```

xrstf commented on hack/ci/run-default-application-e2e-test.sh:
```diff
@@ -0,0 +1,127 @@
+    exit 1
+    ;;
+  esac
+done
```
I think the script will rarely be run by humans, maybe we can skip the entire flag parsing and just use env variables?

iakmc replied:
resolved

xrstf commented on pkg/test/e2e/default_app_catalog_tests/application_test.go:
```diff
@@ -0,0 +1,366 @@
+
+func TestClusters(t *testing.T) {
+	if applicationName == "" || applicationVersion == "" || applicationInstallationName == "" || applicationNamespace == "" || key == "" || names == "" {
+		return
```
If not setting these values is a misconfiguration, this should be a t.Fatal() so we notice and are not just happy that one of our prowjobs finishes really, really quick 😁

xrstf commented on pkg/test/e2e/default_app_catalog_tests/application_test.go:
```diff
@@ -0,0 +1,366 @@
+	testUserCluster(ctx, t, tLogger, client)
+}
+
+//gocyclo:ignore
```
I think the linter is trying to tell us something. Why are just we just ignoring it?

iakmc replied:
Maybe you don't remember, but you showed me the Cilium test as an example and used this function directly from the Cilium test here: https://github.com/kubermatic/kubermatic/blob/f51b80d6a913e522721a91445143b62810b4948d/pkg/test/e2e/cilium/cilium_test.go#L182

xrstf commented on pkg/test/e2e/default_app_catalog_tests/application_test.go:
```diff
@@ -0,0 +1,366 @@
+	tLogger.Info("Running tests...")
+
+	tLogger.Info("Check if ApplicationInstallation exists")
+	err = wait.PollLog(ctx, tLogger, 2*time.Second, 8*time.Minute, func(ctx context.Context) (error, error) {
```
Just out of curiosity, how did you come up with 8 minutes?

iakmc replied:
I can't tell you right now which application it was, but 5 minutes wasn't enough for it, and it was successfully deployed in 8 minutes.

xrstf replied:
Fair enough

embik commented on pkg/test/e2e/default_app_catalog_tests/application_test.go:
```diff
@@ -0,0 +1,365 @@
+	tLogger.Info("Check if ApplicationInstallation exists")
+	err = wait.PollLog(ctx, tLogger, 2*time.Second, 10*time.Minute, func(ctx context.Context) (error, error) {
+		app := &appskubermaticv1.ApplicationInstallation{}
+		if err := client.Get(context.Background(), types.NamespacedName{Namespace: applicationNamespace, Name: applicationInstallationName}, app); err != nil {
```
```suggestion
		if err := client.Get(ctx, types.NamespacedName{Namespace: applicationNamespace, Name: applicationInstallationName}, app); err != nil {
```

embik reviewed (changes requested):
Minor nit, otherwise looking good. Please also address @xrstf's [comment](https://github.com/kubermatic/kubermatic/pull/14312#discussion_r2033846461).

iakmc committed:
This is a combination of 2 commits.

E2E tests for default application catalog

Signed-off-by: Iskren <iskren@kubermatic.com>

download kube test binaries in conformance test ProwJobs (#14324)

download kube test binaries in conformance test ProwJobs (#14324)

e2e tests for the applications from the default app catalog

Signed-off-by: Iskren <iskren@kubermatic.com>

keep the convention for the file names

Signed-off-by: Iskren <iskren@kubermatic.com>

prow jobs added

Signed-off-by: Iskren <iskren@kubermatic.com>

the test is executable

Signed-off-by: Iskren <iskren@kubermatic.com>

add permissions to the file

Signed-off-by: Iskren <iskren@kubermatic.com>

fixed application test

Signed-off-by: Iskren <iskren@kubermatic.com>

remove flag, use env vars

Signed-off-by: Iskren <iskren@kubermatic.com>

resolved Cristoph comment

Signed-off-by: Iskren <iskren@kubermatic.com>

Resolved Christoph's comment

Signed-off-by: Iskren <iskren@kubermatic.com>

fixed Marvin's comment

Signed-off-by: Iskren <iskren@kubermatic.com>

E2E tests for default application catalog

Signed-off-by: Iskren <iskren@kubermatic.com>

fix linter issues

Signed-off-by: Iskren <iskren@kubermatic.com>

fixed package

Signed-off-by: Iskren <iskren@kubermatic.com>

default app catalog

Signed-off-by: Iskren <iskren@kubermatic.com>

fix

Signed-off-by: Iskren <iskren@kubermatic.com>

resolved Simon's comment

Signed-off-by: Iskren <iskren@kubermatic.com>

embik reviewed (approved):
/approve

SimonTheLeg commented on hack/ci/run-default-application-e2e-test.sh:
```diff
@@ -0,0 +1,104 @@
+if [[ -s "$pathToFile" ]]; then
+  versions=$(yq eval '.spec.versions[0].version' "$pathToFile")
+
+  # Extract the defaultValuesBlock using sed (this will capture everything after 'defaultValuesBlock:' until 'documentationURL')
```
nit: 
```suggestion
  # Extract the defaultValuesBlock (this will capture everything after 'defaultValuesBlock:' until 'documentationURL')
```

SimonTheLeg commented on hack/ci/run-default-application-e2e-test.sh:
```diff
@@ -0,0 +1,104 @@
+  # Extract the defaultValuesBlock using sed (this will capture everything after 'defaultValuesBlock:' until 'documentationURL')
+  defaultValuesBlock=$(yq eval '.spec.defaultValuesBlock' "$pathToFile")
+
+  echo "Processing default values block: $defaultValuesBlock"
```
nit
```suggestion
  echodate "Processing default values block: $defaultValuesBlock"
```

SimonTheLeg commented on hack/ci/run-default-application-e2e-test.sh:
```diff
@@ -0,0 +1,104 @@
+  echo "Processing default values block: $defaultValuesBlock"
+
+  for version in $versions; do
+    echo "Processing version: $version"
```
```suggestion
    echodate "Processing version: $version"
```

SimonTheLeg commented on hack/ci/run-default-application-e2e-test.sh:
```diff
@@ -0,0 +1,104 @@
+echodate "File path: $pathToFile"
+
+if [[ -s "$pathToFile" ]]; then
+  versions=$(yq eval '.spec.versions[0].version' "$pathToFile")
```
This will only give you the first item in the array aka the first version.

Which we can also see the test only runs against the first version, not all of them

For example for Argo (should be 3 versions)
![grafik](https://github.com/user-attachments/assets/052d5b4d-dd83-4438-a7fe-cd3068112a62)

I think this should work
```suggestion
  versions=$(yq eval '.spec.versions[].version' "$pathToFile")
```
